File: dsa/queues/ruby/background_job.rb
require 'test/unit'
require 'net/http'
require 'uri'
require 'byebug'

class Node
  attr_accessor :value, :next, :prev

  def initialize(value)
    @value = value
    @next = nil
    @prev = nil
  end
end

class Deque
  attr_reader :head, :tail

  def initialize
    @head = nil
    @tail = nil
  end

  def rpush(value)
    node = Node.new(value)

    if @tail.nil?
      @tail = node
      @head = node
      return
    end

    node.prev = @tail
    @tail.next = node
    @tail = node
  end

  def lpush(value)
    node = Node.new(value)

    if @head.nil?
      @head = node
      @tail = node
      return
    end

    node.next = @head
    @head.prev = node
    @head = node
  end

  def rpop
    return unless @tail

    @tail.value.tap do
      @tail = @tail.prev
      @tail.next = nil if @tail
      @head = nil unless @tail
    end
  end

  def lpop
    return unless @head

    @head.value.tap do
      @head = @head.next
      @head.prev = nil if @head
      @tail = nil unless @head
    end
  end

  def rpoplpush(other)
    rpop.tap do |value|
      other.lpush(value) if value
    end
  end

  # O(n)
  def to_a
    array = []

    pointer = @head

    while pointer
      array << pointer.value

      pointer = pointer.next
    end

    array
  end
end

class BlockingDeque
  def initialize
    @deque = Deque.new
    @mutex = Mutex.new
    @emitter = ConditionVariable.new
  end

  def rpush(value)
    @mutex.synchronize do
      @deque.rpush(value)
      @emitter.signal
    end

    nil
  end

  def lpush(value)
    @mutex.synchronize do
      @deque.lpush(value)
      @emitter.signal
    end

    nil
  end

  def brpop
    @mutex.synchronize do
      @emitter.wait(@mutex) unless @deque.tail

      @deque.rpop
    end
  end

  def blpop
    @mutex.synchronize do
      @emitter.wait(@mutex) unless @deque.head

      @deque.lpop
    end
  end

  def brpoplpush(other)
    brpop.tap do |value|
      other.lpush(value) if value
    end
  end

  def lpop = @deque.lpop
  def rpop = @deque.rpop
  def rpoplpush(other) = @deque.rpoplpush(other)
  def to_a = @deque.to_a
end

class Worker
  MAX_RETRIES = 3

  attr_reader :task_queue, :processing_queue, :dlq

  def initialize(task_queue)
    @task_queue = task_queue
    @processing_queue = BlockingDeque.new
    @dlq = BlockingDeque.new

    @retries = Hash.new(0)
  end

  def start!
    Thread.new do
      wid = Thread.current.object_id
      puts "Worker #{wid} started. Waiting for tasks..."

      loop do
        task = @task_queue.brpoplpush(@processing_queue)

        process(wid, task)
      end
    end
  end

  def process(wid, task)
    puts "[worker-#{wid}] Processing task #{task}..."

    begin
      Net::HTTP.get(URI(task))
      @processing_queue.lpop

      puts "[worker-#{wid}] Task #{task} processed successfully!"
    rescue => e
      puts "Error processing task #{task}: #{e.message}"
      @retries[task] += 1

      if @retries[task] > MAX_RETRIES
        puts "[worker-#{wid}] Max retries reached for task #{task}. Moving to DLQ..."
        @processing_queue.rpoplpush(@dlq)
      else
        puts "[worker-#{wid}] Retrying task #{task}..."
        @processing_queue.rpoplpush(@task_queue)
      end
    end
  end
end

class SidequicoTest < Test::Unit::TestCase
  def test_complex
    task_queue = BlockingDeque.new

    task_queue.lpush('https://hub.dummyapis.com/delay?seconds=1')

    worker = Worker.new(task_queue)

    assert_equal(1, worker.task_queue.to_a.size)

    worker.start!

    sleep 0.1 while worker.processing_queue.to_a.size.zero?
    assert_equal(1, worker.processing_queue.to_a.size)
    assert_equal(0, worker.task_queue.to_a.size)

    sleep 0.1 until worker.processing_queue.to_a.size.zero?
    assert_equal(0, worker.processing_queue.to_a.size)
    assert_equal(0, worker.task_queue.to_a.size)

    task_queue.lpush('https://wroooong.coms')

    sleep 0.1 while worker.dlq.to_a.size.zero?
    assert_equal(1, worker.dlq.to_a.size)
    assert_equal(0, worker.processing_queue.to_a.size)
    assert_equal(0, worker.task_queue.to_a.size)
  end
end

class Sidequico
  def initialize
    @task_queue = BlockingDeque.new
    @worker = Worker.new(@task_queue)
  end

  def run!
    @worker.start!

    loop do
      option = menu

      case option
      when 1
        print 'Digite a URL: '
        url = gets.chomp

        @task_queue.rpush(url)
      when 2
        puts @worker.task_queue.to_a
      when 3
        puts @worker.processing_queue.to_a
      when 4
        puts @worker.dlq.to_a
      when 5
        break
      else
        puts 'Opção inválida'
      end
    end

    puts 'Bye'
  end

  def menu
    sleep 0.05
    puts "====== Sidequico ======"

    puts '## Menu ##'
    puts '#1 - Enfileirar'
    puts '#2 - Ver fila'
    puts '#3 - Ver processamento'
    puts '#4 - Ver DLQ'
    puts '#5 - Sair'
    puts

    print 'Escolha uma opção: '
    gets.to_i
  end
end

#Sidequico.new.run!

File: dsa/queues/rust/background_job.rs
use std::sync::{Arc, Mutex};
use std::sync::Condvar;
use std::thread;
use std::collections::HashMap;
use std::time::Duration;

struct Node<T> {
    value: T,
    next: Option<Arc<Mutex<Node<T>>>>,
    previous: Option<Arc<Mutex<Node<T>>>>,
}

impl<T> Node<T> {
    fn new(value: T) -> Node<T> {
        Node { value, next: None, previous: None, }
    }
}

struct Deque<T> {
    head: Option<Arc<Mutex<Node<T>>>>,
    tail: Option<Arc<Mutex<Node<T>>>>,
}

#[allow(dead_code)]
impl<T> Deque<T> {
    fn new() -> Deque<T> {
        Deque { head: None, tail: None }
    }

    fn lpush(&mut self, value: T) {
        let node = Arc::new(Mutex::new(Node::new(value)));

        match &self.head {
            Some(head_ref) => {
                node.lock().unwrap().next = Some(head_ref.clone());
                head_ref.lock().unwrap().previous = Some(node.clone());
                self.head = Some(node.clone());
            },
            None => {
                self.head = Some(node.clone());
                self.tail = Some(node.clone());
            }
        }
    }

    fn rpush(&mut self, value: T) {
        let node = Arc::new(Mutex::new(Node::new(value)));

        match &self.tail {
            Some(tail_ref) => {
                node.lock().unwrap().previous = Some(tail_ref.clone());
                tail_ref.lock().unwrap().next = Some(node.clone());
                self.tail = Some(node.clone());
            },
            None => {
                self.head = Some(node.clone());
                self.tail = Some(node.clone());
            }
        }
    }

    fn lpop(&mut self) -> Option<T> where T: Clone {
        match &self.head {
            Some(head_ref) => {
                let value = head_ref.lock().unwrap().value.clone();
                let next_ref = head_ref.lock().unwrap().next.clone();

                match &next_ref {
                    Some(next_ref) => {
                        next_ref.lock().unwrap().previous = None;
                        self.head = Some(next_ref.clone());
                    },
                    None => {
                        self.head = None;
                        self.tail = None;
                    }
                }
                Some(value)
            },
            None => None
        }
    }

    fn rpop(&mut self) -> Option<T> where T: Clone {
        match &self.tail {
            Some(tail_ref) => {
                let value = tail_ref.lock().unwrap().value.clone();
                let previous_ref = tail_ref.lock().unwrap().previous.clone();

                match &previous_ref {
                    Some(previous_ref) => {
                        previous_ref.lock().unwrap().next = None;
                        self.tail = Some(previous_ref.clone());
                    },
                    None => {
                        self.head = None;
                        self.tail = None;
                    }
                }
                Some(value)
            },
            None => None
        }
    }

    fn rpoplpush(&mut self, other: &mut Deque<T>) -> Option<T> where T: Clone {
        let value = self.rpop();

        match &value {
            Some(value) => other.lpush(value.clone()),
            None => ()
        }

        value
    }

    fn to_vec(&self) -> Vec<T> where T: Clone {
        let mut vec = Vec::new();
        let mut current = self.head.clone();

        while let Some(node) = current {
            vec.push(node.lock().unwrap().value.clone());
            current = node.lock().unwrap().next.clone();
        }

        vec
    }
}

struct BlockingDeque<T> {
    store: Mutex<Deque<T>>,
    emitter: Condvar,
}

#[allow(dead_code)]
impl<T> BlockingDeque<T> {
    fn new() -> BlockingDeque<T> {
        BlockingDeque {
            store: Mutex::new(Deque::new()),
            emitter: Condvar::new(),
        }
    }

    fn lpush(&self, t: T) {
        self.store.lock().unwrap().lpush(t);
        self.emitter.notify_one();
    }

    fn rpush(&self, t: T) {
        self.store.lock().unwrap().rpush(t);
        self.emitter.notify_one();
    }

    fn blpop(&self) -> Option<T> where T: Clone {
        let mut store = self.store.lock().unwrap();

        while store.head.is_none() {
            store = self.emitter.wait(store).unwrap();
        }

        store.lpop()
    }

    fn brpop(&self) -> Option<T> where T: Clone {
        let mut store = self.store.lock().unwrap();

        while store.head.is_none() {
            store = self.emitter.wait(store).unwrap();
        }

        store.rpop()
    }

    fn brpoplpush(&self, other: &BlockingDeque<T>) -> Option<T> where T: Clone {
        let mut store = self.store.lock().unwrap();

        while store.tail.is_none() {
            store = self.emitter.wait(store).unwrap();
        }

        let value = store.rpoplpush(&mut other.store.lock().unwrap());

        self.emitter.notify_one();

        value
    }

    fn lpop(&self) -> Option<T> where T: Clone {
        self.store.lock().unwrap().lpop()
    }

    fn rpop(&self) -> Option<T> where T: Clone {
        self.store.lock().unwrap().rpop()
    }

    fn to_vec(&self) -> Vec<T> where T: Clone {
        self.store.lock().unwrap().to_vec()
    }
}

struct Worker {
    task_queue: Arc<BlockingDeque<i32>>,
    processing_queue: Arc<BlockingDeque<i32>>,
    dlq: Arc<BlockingDeque<i32>>,
    retries: Arc<Mutex<HashMap<i32, i32>>>,
}

impl Worker {
    fn new() -> Worker {
        Worker {
            task_queue: Arc::new(BlockingDeque::new()),
            processing_queue: Arc::new(BlockingDeque::new()),
            dlq: Arc::new(BlockingDeque::new()),
            retries: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

struct Processor {}

impl Processor {
    fn start(worker: Arc<Worker>) {
        let task_queue = worker.task_queue.clone();
        let processing_queue = worker.processing_queue.clone();
        let dlq = worker.dlq.clone();
        let retries = worker.retries.clone();

        thread::spawn(move || {
            loop {
                let task = task_queue.brpoplpush(&processing_queue);

                match task {
                    Some(task) => {
                        if task == 42 {
                            let mut retries = retries.lock().unwrap();
                            let current_retry = *retries.entry(task).or_insert(0) + 1;
                            retries.insert(task, current_retry);

                            if current_retry > 3 {
                                processing_queue.brpoplpush(&dlq);
                            } else {
                                processing_queue.brpoplpush(&task_queue);
                            }
                        } else {
                            let duration = Duration::from_secs(task as u64);
                            thread::sleep(duration);
                            processing_queue.brpop();
                        }
                    },
                    None => ()
                }
            }
        });
    }
}

mod tests {
    use super::*;

    #[test]
    fn test_background_processor() {
        // Start a worker
        let worker = Arc::new(Worker::new());
        Processor::start(worker.clone());

        // Push a task to the task_queue
        worker.task_queue.lpush(1);
        assert_eq!(worker.task_queue.to_vec().len(), 1);
        assert_eq!(worker.processing_queue.to_vec().len(), 0);

        // Wait for the task to be processed
        while worker.processing_queue.to_vec().len() == 0 {
            thread::sleep(Duration::from_millis(100));
        }
        assert_eq!(worker.task_queue.to_vec().len(), 0);
        assert_eq!(worker.processing_queue.to_vec().len(), 1);

        // Push a task to the task_queue (a task that will fail)
        worker.task_queue.lpush(42);

        // Wait for the task to be processed, failed and put in the DLQ
        while worker.dlq.to_vec().len() == 0 {
            thread::sleep(Duration::from_millis(100));
        }
        assert_eq!(worker.task_queue.to_vec().len(), 0);
        assert_eq!(worker.processing_queue.to_vec().len(), 0);
        assert_eq!(worker.dlq.to_vec().len(), 1);
    }
}
